flang/lib/Lower/IO.cpp

```diff
@@ -196,6 +196,87 @@ static void makeNextConditionalOn(Fortran::lower::FirOpBuilder &builder,
   builder.setInsertionPointToStart(&ifOp.thenRegion().front());
 }
 
+/// Generate a namelist I/O call.  The data structure containing the namelist
+/// group information is defined in runtime header file namelist.h as:
+///
+///     class NamelistGroup {
+///     public:
+///       struct Item {
+///         const char *name; // NUL-terminated lower-case
+///         const Descriptor &descriptor;
+///       };
+///       const char *groupName; // NUL-terminated lower-case
+///       std::size_t items;
+///       const Item *item; // in original declaration order
+///     };
+static void genNamelistIO(Fortran::lower::AbstractConverter &converter,
+                          mlir::Value cookie, mlir::FuncOp funcOp,
+                          Fortran::semantics::Symbol &symbol, bool checkResult,
+                          mlir::Value &ok,
+                          Fortran::lower::StatementContext &stmtCtx) {
+  const auto &details = symbol.get<Fortran::semantics::NamelistDetails>();
+  auto &builder = converter.getFirOpBuilder();
+  auto loc = converter.getCurrentLocation();
+  auto *context = builder.getContext();
+  auto idxTy = builder.getIndexType();
+  auto sizeTy = builder.getIntegerType(8 * sizeof(std::size_t));
+  auto charRefTy = fir::ReferenceType::get(builder.getIntegerType(8));
+  auto descRefTy =
+      fir::ReferenceType::get(fir::BoxType::get(mlir::NoneType::get(context)));
+  auto listTy = fir::SequenceType::get(
+      details.objects().size(),
+      mlir::TupleType::get(context, {charRefTy, descRefTy}));
+  auto groupTy = mlir::TupleType::get(
+      context, {charRefTy, sizeTy, fir::ReferenceType::get(listTy)});
+  auto stringAddress = [&](const Fortran::semantics::Symbol &symbol) {
+    auto global =
+        createStringLiteral(builder, loc, symbol.name().ToString() + '\0');
+    return builder.createConvert(loc, charRefTy, fir::getBase(global));
+  };
+  makeNextConditionalOn(builder, loc, checkResult, ok);
+  auto zero = builder.createIntegerConstant(loc, idxTy, 0);
+  auto one = builder.createIntegerConstant(loc, idxTy, 1);
+  auto two = builder.createIntegerConstant(loc, idxTy, 2);
+  auto listAddr = builder.create<fir::AllocaOp>(loc, listTy);
+  mlir::Value list = builder.create<fir::LoadOp>(loc, listAddr);
```

Comment:
Given you are not relying on some data already being set in the storage like you did with the names in the global, it would probably be best to start with an undefined (I think this will avoid LLVM optimizers having to think whether or not the storage can change between the load and final store , if it cannot be sure, it would probably generate another temp from the load, modify it, and copy it back).
```
mlir::Value list = builder.create<fir::UndefOp>(loc, fir::dyn_cast_ptrEleTy(listAddr.getType()));
```

Reply:
done; thanks